tokio-sync/src/oneshot.rs

```diff
@@ -0,0 +1,509 @@
+//! A channel for sending a single message between asynchronous tasks.
+
+use loom::{
+    futures::task::{self, Task},
+    sync::CausalCell,
+    sync::atomic::AtomicUsize,
+};
+
+use futures::{Async, Future, Poll};
+
+use std::fmt;
+use std::mem::{self, ManuallyDrop};
+use std::sync::Arc;
+use std::sync::atomic::Ordering::{self, Acquire, AcqRel};
+
+/// Sends a value to the associated `Receiver`.
+///
+/// Instances are created by the [`channel`](fn.channel.html) function.
+#[derive(Debug)]
+pub struct Sender<T> {
+    inner: Option<Arc<Inner<T>>>,
+}
+
+/// Receive a value from the associated `Sender`.
+///
+/// Instances are created by the [`channel`](fn.channel.html) function.
+#[derive(Debug)]
+pub struct Receiver<T> {
+    inner: Option<Arc<Inner<T>>>,
+}
+
+pub mod error {
+    //! Oneshot error types
+
+    /// Error returned by the `Future` implementation for `Receiver`.
+    #[derive(Debug)]
+    pub struct RecvError(pub(super) ());
+
+    /// Error returned by the `try_recv` function on `Receiver`.
+    #[derive(Debug)]
+    pub struct TryRecvError(pub(super) ());
```

Comment:
It'd be nice if `TryRecvError` could tell whether the channel is closed or just empty.

Reply:
Forward compat change!